gcp_variant_transforms/libs/variant_partition.py

```diff
@@ -34,26 +43,216 @@
+  If self._interval_tree != None this instance is encoding multiple regions.
+  """
 
   def __init__(self):
-    if _DEFAULT_NUM_PARTITIONS <= _RESERVED_AUTO_PARTITIONS:
+    # Each instance contains either:
+    #    one whole chromosome (in that case _index != -1).
+    self._index = -1
+    #    OR multiple regions of one chromosome (_interval_tree != None).
+    self._interval_tree = None
+
+  def add_interval(self, start, end, index):
+    if self._index != -1:
+      raise ValueError('Can not add region to an existing full chromosome.')
+    if start < 0:
+      raise ValueError('Start position on a region cannot be negative: {}'.
+                       format(start))
+    if end <= start:
+      raise ValueError('End position must be larger than start position: {} '
+                       'vs {}'.format(end, start))
+    if index < 0:
+      raise ValueError('Index of a region cannot be negative {}'.format(index))
+    if not self._interval_tree:
+      # interval tree has not been initiated.
+      self._interval_tree = intervaltree.IntervalTree()
+    else:
+      # Check the existing interval tree for possible overlap with new interval.
+      if self._interval_tree.overlaps_range(start, end):
+        raise ValueError('Cannot add overlapping region {}-{}'.
+                         format(start, end))
+    # If everything goes well we add the new region to the interval tree.
+    self._interval_tree.addi(start, end, index)
+
+  def add_index(self, index):
+    if self._interval_tree is not None:
+      raise ValueError('Can not add full chromosome to existing multi region.')
+    if self._index != -1:
+      raise ValueError('This instance has already been assinged an index.')
+    if index < 0:
+      raise ValueError('Index of a region cannot be negative {}'.format(index))
+    self._index = index
+
+  def get_index(self, pos=0):
+    if self._interval_tree is None and self._index == -1:
+      raise ValueError('This instance has not been assigned any index.')
+    if self._interval_tree is None:
+      return self._index
+    else:
+      matched_intervals = self._interval_tree.search(pos)
+      if len(matched_intervals) > 1:
+        raise ValueError('Search for pos {} returned more than 1 result'.
+                         format(pos))
+      if len(matched_intervals) == 1:
+        return next(iter(matched_intervals)).data
+      else:
+        return -1
+
+  @property
+  def is_full_chromosome(self):
+    return self._index != -1
+
+class VariantPartition(object):
+  """Partition variants based on their reference_name and/or position.
+
+  This class has 2 operating modes:
+    1) No config file is given (config_file_path == ''):
+        Automatically partition variants based on their reference_name.
+    2) A config file is given:
+        partitions variants based on input config file.
+  """
+
+  def __init__(self, config_file_path=''):
+    if not config_file_path:
```

Comment:
as previously discussed: it's much cleaner to have a 'state' with all of the parameters initialized to default values (e.g. `self._num_partitions = _DEFAULT_NUM_PARTITIONS`, `self._residual_partition_index = -1`, etc). Then, if config file is given, you'd update the state accordingly. It also makes the code much easier to follow, as currently it's not clear what is the state of the class when a config file is given (e.g. how is `_by_ref_name` different than `_reference_name_to_partition_map`? are they both really needed?).